### ISSUE TITLE
cryptographically secure keyed rolling hash / MAC?

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#2128]
is there such a thing? could we use it to chunk and compute the storage key in one go?

currently the chunker computes a rolling hash (buzhash) over a 4095 bytes window to determine cutting places from the content. the rolling hash value is thrown away then and a cryptographic hash ([hmac-]sha256, blake2b) is computed for the storage key / for deduplication.

if we only used a CSKRH to save one step in processing, the window size would grow to be equal to the chunk size (we can't have small window and big chunk, that would cause hash collisions) and also the output size would need to be >= 256bits.

**Comment by enkore** [comment]
[StackOverflow](http://crypto.stackexchange.com/a/16087/8351)

Update: looks like this now redirects to [StackExchange](https://crypto.stackexchange.com/questions/16082/cryptographically-secure-keyed-rolling-hash-function/16087#16087)

The suggested scheme looks correct to me, however, it's performance is inferior to what we have right now, even with hardware-accelerated AES.

I think if we want to get more performance there is a simple tradeoff; don't cut on byte boundaries. Cut on aligned 4/8 byte boundaries, and process 4/8 bytes in each step (I did some experiments in this direction called beer4hash and beer8hash, each was >1 GB/s, otherwise same construction as buzhash).

If we want to make it more secure (ie. less leaky)... ahem. There is just not a whole lot of bits there to begin with, and many files will be only 1-2 chunks.

**Comment by ThomasWaldmann** [comment]
Performance is only inferior if R is slower. But likely it would have to be due to the required properties. E could be done only when needed (at cutting places).

Can you link to the beerNhash code?

**Comment by enkore** [comment]
> Performance is only inferior if R is slower. But likely it would have to be due to the required properties. E could be done only when needed (at cutting places).

No, you can't -- if you want what's in the topic you have to cut according to the bits after encryption, meaning that for byte-granularity you have to encrypt the current hash after every byte. A high-end Haswell PC might do something like 100 - 200 million AES blocks per second, which would then be about the throughput to expect. On lower end platforms this will look much worse.

If you don't need that, then there is not much point to using that or a similar construction.

I don't think I pushed/cleaned them up yet but it's pretty straightforward stuff.

**Comment by enkore** [comment]
Also a problem: It's still just a 32 bit hash. The block cipher is only used as a permutation 32b -> 128b (w/ AES 128b block size), but that doesn't change anything about there only being 2**32 possible outputs, which is far too weak for using that construction.

I think a CSKRH is even a bit of an oxymoron: the RH properties we want means that it's linear over short byte ranges (and then permuted / noise injected through table like in buzhash), but that's something we *don't* want for the ID hash.

**Comment by ThomasWaldmann** [comment]
I didn't suggest using the current 32bit RH as basis, so we would just need a much longer RH.
To be used as id hash, it would also need to cover the whole chunk (e.g. 2MiB rather than 4095B).

**Comment by anarcat** [comment]
on restic's side, they use the Rabin Fingerprinting algorithm, which is (according to Wikipedia), possibly slower than the buzhash used by Borg. Yet when I asked @fd0 about this, he said this wasn't the bottleneck so I wonder if it's the case that we really need to optimize this for borg or that would be misplaced optimization...

**Comment by fd0** [comment]
For the reference, the chunker is written in pure Go (no fancy assembler optimizations or so), uses 64 bit Rabin fingerprints as a rolling hash, and on my machine (`Intel(R) Core(TM) i5-6440HQ CPU @ 2.60GHz`) comes to 391MB/s:
```
$ go test -bench .
BenchmarkChunker-4                    20          85682339 ns/op         391.61 MB/s
```

Overall, that's more than enough for restic, the other stages that we have (most importantly the SHA256 for identifying a plaintext chunk and the SHA256 for the storage file name over the encrypted content) are the slowdown here.

**Comment by enkore** [comment]
Isn't restic's data processing parallelized, though?

**Comment by anarcat** [comment]
so I guess there *would* be an advantage in changing the rolling hash in borg, considering it's more of a serial pipeline right now. yet this probably means all chunks will change with a new algorithm, which might be too disruptive to justify such a change. best would be to parallelize borg, really... and that's another issue (#37) but at least it wouldn't affect the repository format.

thanks for the feedback @fd0 and @enkore ! :)

**Comment by espadrine** [comment]
I don’t see the value in layering the PRP on top of the ε-universal rolling hash.

You can cut the input file in 128-bit blocks, they can each be encrypted by AES-128 concurrently, and you cut after every block whose 128-bit ciphertext number is below a given value (set to obtain an average frequency).

As the SO answer rightly mentions, a PRP is a PRF, and all you care about to prove security is that the output is a PRF.

**Comment by ThomasWaldmann** [comment]
I think I have something now in #8903, please review!

**Comment by ThomasWaldmann** [comment]
@espadrine What you said was a bit above my head. I'm not a mathematician nor a cryptographer, just a interested nerd (and I did take a crypto course some years ago). 

But I tried to understand what you said and if I understood it correctly, that construction has the problem that e.g. block contents at the beginning of the file influence the value of all Ci after it. Changing the contents of that block also changes all Ci after it.

But for a chunker, the chunking decision should be only based on the windows content and the cutting point should not change if a block from far before the window changes.

**Comment by ThomasWaldmann** [comment]
Closing this as answered.

**Short answer: no — and not because nobody has built one yet.** The two requirements contradict each other. A cut decision must depend *only* on a bounded window, otherwise chunk boundaries never resync after an edit and dedup collapses. A chunk ID must depend on *every* byte of the chunk. No single per-byte primitive can do both, so the "save one step" idea from my original post is not realizable. @enkore's "a CSKRH is even a bit of an oxymoron" comment above holds up.

Measured out of curiosity (Apple M3 Pro, single thread, standalone C, AES via ARM crypto extensions, 256 MiB):

| construction | throughput |
| --- | --- |
| gear rolling hash | 1700 MB/s |
| buzhash64 (standalone reimplementation) | 850 MB/s |
| Rabin, table-driven (a real universal hash) | 491 MB/s |
| gear + AES-128 per byte (cut decision only) | 985 MB/s |
| Rabin + AES-128 per byte | 492 MB/s |
| gear + AES per byte + chained AES chunk tag ("both in one pass") | **122 MB/s** |

For reference, HMAC-SHA256 runs at ~2600 MB/s on the same machine, so what borg does today — chunker, then ID hash, in series — is ~640 MB/s. The "one pass for both" construction is ~5x *slower*, because the ID accumulator is a serial AES dependency chain that nothing can pipeline away. The idea is not merely blocked in theory, it is counterproductive in practice.

**The other half of the question does have an answer now.** A keyed chunker whose cut points provably leak nothing is exactly what ["Breaking and Fixing Content-Defined Chunking"](https://eprint.iacr.org/2025/558) (Truong, Merz, Scarlata, Günther, Paterson, 2025) constructs: a rolling universal hash over the window, followed by a block cipher, cutting on the PRF output — one AES call per byte. That is a chunker only; it does not give you the chunk ID, and its window state is deliberately *not* collision resistant.

Worth noting from the numbers above: the AES is not the expensive part. On the Rabin hash it is free (491 → 492 MB/s, the dependent table load already owns the critical path); on gear it costs 42% and still lands above our current buzhash64. The real cost driver is which universal hash sits underneath, which is an engineering question rather than a crypto one.

One dead end worth recording, because it looks attractive: a cheap keyed pre-filter that only invokes the PRF on filter hits does **not** work. Observed cut points are a subset of filter hits, so each cut still yields equations in the filter key; once that is recovered, an attacker can test a candidate file by checking that every observed cut lands on a filter hit. That is a fingerprinting distinguisher even though the cut decisions themselves stay unpredictable. The PRF has to gate every position.

Further chunker work continues in #8841 and #9280, and the current status of the 2025 CDC attacks against borg is tracked at https://github.com/borgbackup/borg/wiki/CDC-issues-reported-2025.
